src/common/eeprom.cpp

```diff
@@ -623,120 +627,127 @@ static int eeprom_convert_from_v2(void) {
+    eevars = vars;
 }
 
-// conversion function for old version 9 (v 4.3.2)
-static int eeprom_convert_from_v9(void) {
+/**
+ * @brief conversion function for old version 9 (v 4.3.2)
+ *
+ * @param eevars input/output eeprom struct
+ */
+static void eeprom_convert_from_v9(eeprom_vars_t &eevars) {
     eeprom_vars_t vars = eeprom_var_defaults;
     eeprom_init_FW_identifiers(vars);
 
-    eeprom_import_block(EEVAR_FILAMENT_TYPE, EEVAR_DEVHASH_IN_QR, &(vars.FILAMENT_TYPE));
-
-    eeprom_make_patches(vars);
-
-    eeprom_save_upgraded(vars);
-    return 1;
-}
+    eeprom_import_block(EEVAR_FILAMENT_TYPE, EEVAR_DEVHASH_IN_QR, vars, eevars);
+    eevars = vars;
+}
+
+/**
+ * @brief conversion function for new version format (features, firmware version/build)
+ * does not change crc, it is changed automatically by write function
+ *
+ * @param eevars eeprom struct in RAM
+ * @return true updated (changed)
+ * @return false not changed
+ */
+static bool eeprom_convert_from(eeprom_vars_t &eevars) {
+    switch (eevars.VERSION) {
+    case 2:
+        eeprom_convert_from_v2(eevars);
+        break;
+    case 4:
+        eeprom_convert_from_v4(eevars);
+        break;
+    case 6:
+        eeprom_convert_from_v6(eevars);
+        break;
+    case 7:
+        eeprom_convert_from_v7(eevars);
+        break;
+    case 8:
+        eeprom_convert_from_v8(eevars);
+        break;
+    case 9:
+        eeprom_convert_from_v9(eevars);
+        break;
+    default:
+        return false;
+    }
 
-// conversion function for new version format (features, firmware version/build)
-static int eeprom_convert_from(uint16_t version, uint16_t features) {
-    if (version == 2)
-        return eeprom_convert_from_v2();
-    if (version == 4)
-        return eeprom_convert_from_v4();
-    if (version == 6)
-        return eeprom_convert_from_v6();
-    if (version == 7)
-        return eeprom_convert_from_v7();
-    if (version == 8)
-        return eeprom_convert_from_v8();
-    if (version == 9)
-        return eeprom_convert_from_v9();
-    return 0;
+    eeprom_make_patches(eevars);
+    return true;
 }
 
 // version independent crc32 check
-static int eeprom_check_crc32(void) {
-    uint16_t datasize;
-    uint32_t crc;
-    datasize = variant8_get_ui16(eeprom_get_var(EEVAR_DATASIZE));
-    if (datasize > EEPROM_MAX_DATASIZE)
-        return 0;
-    st25dv64k_user_read_bytes(EEPROM_ADDRESS + datasize - 4, &crc, 4);
+static bool eeprom_check_crc32(const eeprom_vars_t &eevars) {
+    if (eevars.DATASIZE > EEPROM_MAX_DATASIZE)
+        return false;
 #if 1 //simple method
-    uint8_t data[EEPROM_MAX_DATASIZE];
-    uint32_t crc2;
-    st25dv64k_user_read_bytes(EEPROM_ADDRESS, data, datasize);
-    crc2 = crc32_eeprom((uint32_t *)data, (datasize - 4) / 4);
-    return (crc == crc2) ? 1 : 0;
+    uint32_t crc = crc32_eeprom((uint32_t *)(&eevars), (eevars.DATASIZE - 4) / 4);
+    return eevars.CRC32 == crc;
 #else //
 #endif
```

Comment:
Maybe you can remove
```
#if 1
#else //
#endif
```

Reply:
it corresponds with some code in private and would create conflict